File: 1.arrays_and_strings/2. merge_strings_alternately.py
class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        res = []
        n_min = min(len(word1), len(word2))
        
        # Merge overlapping parts
        for i in range(n_min):
            res.append(word1[i])
            res.append(word2[i])
            
        # Append remaining tails (one will be empty)
        res.append(word1[n_min:])
        res.append(word2[n_min:])
        
        return "".join(res)

# Test case
if __name__ == "__main__":
    solution = Solution()
    print(solution.mergeAlternately("abc", "pqr"))    # Output: apbqcr
    print(solution.mergeAlternately("ab", "pqrs"))   # Output: apbqrs
    print(solution.mergeAlternately("abcd", "pq"))   # Output: apbqcd

File: 1.arrays_and_strings/2. merge_strings_alternately.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {string}
 */
var mergeAlternately = function (word1, word2) {
    let result = '';
    const length = Math.max(word1.length, word2.length);

    for (let i = 0; i < length; i++) {
        // Checking for truthiness handles characters and undefined
        if (word1[i]) result += word1[i];
        if (word2[i]) result += word2[i];
    }

    return result;
};

// Test case
console.log(mergeAlternately("abc", "pqr"));    // Output: apbqcr
console.log(mergeAlternately("ab", "pqrs"));   // Output: apbqrs
console.log(mergeAlternately("abcd", "pq"));   // Output: apbqcd
